Adding Rupee Color randomization

AntiBlueQuirk opened the pull request:
I added Rupee color randomization. This can affect any of the six rupee types (1, 5, 20, 50, 200, Silver), and affects both their in-world pickup model, as well as the getitem model (when you get one out of a chest).

Note that there is no logic right now to prevent the randomizer from picking similar, or even identical colors for different denominations, so telling rupees apart could become quite difficult.

Also, I haven't changed any of the pickup messages, simply because I haven't figured out the best way to do that. So the game still refers to 5 Rupees as Blue Rupees, even if they're a different color.

![tst_legend_of_zelda-cmb2](https://user-images.githubusercontent.com/25140955/57813508-f5a95c80-7735-11e9-87ce-b232d4d4c658.png)

totenphal commented:
Took me some time to get this going:

- Gui Implementation is missing. (New Gui is on the way, might not be needed)
- If Silver rupee is Silver Rupee, exception occurs.
- Lighting effect of freestanding models looks weird (This case Indigo on 50 rupees)

![Screenshot_2](https://user-images.githubusercontent.com/11833572/57868006-7766bc00-7802-11e9-9a76-55079bc1b0ab.png)
![Screenshot_3](https://user-images.githubusercontent.com/11833572/57868034-80578d80-7802-11e9-9822-9dc2808c35bc.png)

AntiBlueQuirk commented:
Fixed the crash. That one sneaked its way in recently when I changed the logic around a bit. The GUI just slipped my mind since I've been working on the command line.

As for the lighting bug, I'm not sure exactly how to fix it, or even exactly what is causing the problem. I noticed it during testing, but my first reaction was that it was an emulator bug. I need to do some more testing to track it down, but it definitely has something to do with choice of color.

To render GI (freestanding) rupees, the game runs these display lists:

 - Inner Setup
 - Inner Render
 - Outer Setup
 - Outer Render 

The Inner/Outer Render DL's are common to every rupee type, the Setup DL's are what make each rupee look unique. But literally the only thing they do is load an environment color, and a geometry color. The only bytes I touch in this process are the ones that load those colors, so it must be something to do with how the shader is handling the color.

If I had to guess, the color math is doing an additive blend, and overflowing. This is avoided in the vanilla game because they carefully selected the colors. The randomizer picks those colors... less carefully.

AntiBlueQuirk commented:
Okay, well, after tinkering with it quite bit, I wasn't really able to figure out what the problem with the blend was. It's definitely an underflow/overflow problem, but I can't quite wrap my mind around how they picked the color values they did.

I did quite a bit of digging for this, including building a simple display list disassembler, and figured out that the color math for the outer layer works something like this:

    stage0: (tex - prim) * (lod_frac + tex)
    stage1: (prim - env) * (stage0 + env)

Where `prim` is the third color I override, and `env` is the fourth.

I *did* figure out that if one of those colors has a non-zero channel, the other color needs to have FF in that channel, or the lighting goes weird. So I fixed it by just making color 3 always be #FFFFFF. This isn't a perfect fix, but it still looks pretty good, and allows full control of color 4.

Sorry if all this information seems a bit unnecessary, I just felt like I needed to write it down somewhere.

rattus128 commented:
This actually has a non-cosmetic gameplay impact. If you are yet to loot a 20 or 200, and you see a freestander you don't know and that could be a big, especially in shopsanity.

AntiBlueQuirk commented:
Sure, but the same thing could be said of tunic colors. Like the rest of the cosmetic options, this doesn't have any gameplay impact *besides* messing with your understanding of the game. It defaults to no change anyway, so one could easily just *not* turn on randomization for some or any rupee types.

flagrama commented:
Tunic colors only affect how they look on your body. You will know which one you have equipped because you equipped it. Rupees you don't know anything about until you pick them up even though you can see them almost anywhere.

TestRunnerSRL commented:
Hmm, I'm not sure about this feature. I don't think this really adds much to the rando (especially since you should be able to just check the cosmetic log to see the results). People mentioned it could have possible repercussions on gameplay, but tbh, you're probably getting most your money from fixed locations anyways. Also things like Huge rupees are bigger so the color just makes things slightly more annoying to distinguish even if you aren't aware of the colors.

That being said this is a pretty minor cosmetic change. I was mostly ok with cosmetics to Link and the HUD/System because you interact with those pretty much continuously. But not really sure that applies to rupees. I'm pretty concerned this is going to lead to people wanting to add even more minor cosmetic changes that will lead to a large amount of cosmetics. This one is kinda on the fence for me mainly because you do interact with rupees somewhat frequently, but I'm not really sure this one adds any real value to the play.

AntiBlueQuirk commented:
Fair enough. I mostly just added it because I thought it would be fun. I can definitely see the arguments against it. I won't be upset if this doesn't get merged, I still enjoyed the challenge of getting it working. (And learned a lot about OOT in the process. :smile:)

TestRunnerSRL commented:
I think that's a great mentality to have. I'm probably just going to pass on this for the reasons I mentioned above and I'm kinda hesistent to add too many options in general. It's a good idea tho, but I just don't think it's best fit for this repo.
